src/rust/cryptography-x509-verification/src/types.rs

```diff
@@ -348,10 +367,45 @@ impl<'a> RFC822Name<'a> {
     }
 }
 
+/// An `RFC822Constraint` represents a Name Constraint on email addresses.
+pub enum RFC822Constraint<'a> {
+    /// A constraint for an exact match on a specific email address.
+    Exact(RFC822Name<'a>),
+    /// A constraint for any mailbox on a particular domain.
+    OnDomain(DNSName<'a>),
+    /// A constraint for any mailbox *within* a particular domain.
+    /// For example, `InDomain("example.com")` will match `foo@bar.example.com`
+    /// but not `foo@example.com`, since `bar.example.com` is in `example.com`
+    /// but `example.com` is not within itself.
+    InDomain(DNSName<'a>),
+}
+
+impl<'a> RFC822Constraint<'a> {
+    pub fn new(constraint: &'a str) -> Option<Self> {
+        if let Some(constraint) = constraint.strip_prefix('.') {
+            DNSName::new(constraint).map(Self::InDomain)
```

Comment:
Please don't use map for this stuff, it's just easier to read: `Self::InDomain(DNSName::new(constraint)?)`